annotations and sorting for activities-completing-read

jdtsmith opened the pull request:
(updated) This provides custom annotation and sorting functions for `activities-completing-read`. It shows the number of buffers (and files), for the last or default state, a magit-style color-coded _activity age_, and flags for being active or having a modified set of buffers. 

This is shown below with `vertico` but it should support any UI which handles annotation-functions (most do).   For theme friendliness, it creates no new faces, instead ~borrowing the `vc-annotate` color palette~ building its own color palette (along with `success` and `warning` faces).  Happy to take any input/ideas.

<img width="375" alt="image" src="https://github.com/user-attachments/assets/2ef0c4d7-0113-4c79-8083-b339ee87265a">

alphapapa commented on activities.el:
```diff
@@ -801,6 +819,90 @@ activity's name is NAME."
+		 (window-state (activities-activity-state-window-state state))
+		 (buffers (window-state-buffers window-state))
+		 (files ( activities--map-window-state-leafs window-state
+			  (lambda (l)
```
Nitpick: Please don't use single-letter variable names.  For clarity, just use the word.

alphapapa commented on activities.el:
```diff
@@ -801,6 +819,90 @@ activity's name is NAME."
+		  (list (length buffers)
+			(length (delq nil files))
+			(float-time (time-since time))))))
+	(pcase-let* ((`(,lastb ,lastf ,lastt) (map-elt data 'last))
+		     (`(,defb ,deff ,deft) (map-elt data 'default))
+		     (age (min lastt deft))
+		     (ann (format "%s:%s|%s %s:%s|%s "
+				  (propertize "buffers" 'face 'bold)
+				  (propertize (format "%2d" lastb) 'face 'success)
+				  (propertize (format "%-2d" defb) 'face 'warning)
+				  (propertize "files" 'face 'bold)
+				  (propertize (format "%2d" lastf) 'face 'success)
+				  (propertize (format "%-2d" deff) 'face 'warning)))
```
This looks nice, but the variable names are a bit unclear with those abbreviations.  I'd rather the code be clear and unambiguous, even if it means the formatting suffers.  So please just use the full words in the variable names.

alphapapa commented on activities.el:
```diff
@@ -801,6 +819,90 @@ activity's name is NAME."
+		      (propertize (apply #'format "[%d %s]" (activities--age age)) 'face
+				  `( :foreground ,age-color
+				     :background ,vc-annotate-background))))
+	  (when (< (length age-ann) age-len) ; left-pad
```
Is it necessary to pad this way?  `format` should be able to handle that.

jdtsmith replied:
Yes indeed, doh.

alphapapa reviewed (changes requested):
Thanks for working on this.  Here are a few comments.

jdtsmith commented on activities.el:
```diff
@@ -801,6 +819,90 @@ activity's name is NAME."
+  (cl-loop for (_name . act) in activities maximize
+	   (cl-loop
+	    for type in '(default last)
+	    for func = (intern (format "activities-activity-%s" type))
```
Forgot about that one, thanks.

jdtsmith commented on activities.el:
```diff
@@ -651,6 +652,23 @@ activity's name is NAME."
+			 (push (funcall func (cl-subseq state leaf-pos)) ret)
+		       (dolist (s state) (map-leafs s func)))))))
+      (map-leafs state func))
+    (nreverse ret)))
```
You mean branches in the `pcase`, yes?  One of those is necessary to "do nothing" on "uninteresting lists" that may be present.  

BTW, this revealed a bug in your `pcase` formulation (which I borrowed from `bufferize-window-state`): `(pred list)` is always true.  I think `list` is not the predicate you need; I used `proper-list-p`.

alphapapa replied:
You're right that it was a mistake: `list` is not a predicate.  `listp` is what I should have used.  It may be that it made no difference in practice.  

I think using `proper-list-p` probably isn't necessary over `listp`; the values in the window state object are well-defined, and misusing `list` was apparently working.

alphapapa replied:
Oh, I see how you removed the other clause, so now using `proper-list-p` is likely necessary.

jdtsmith replied:
It made no difference I think because all other cases are caught by your prior cases, so `t` would have even sufficed.   Probably best for safety to use `listp`.  I'm using `proper-list-p` because I need to exclude cons cells like `(min-width . 8)` which I no longer match affirmatively.  I could restore that "do-nothing" case or just insist on a "real list"; happy with whatever would make more sense to you down the road.  `cl-position` does not take kindly to non-nil-terminated cons cells.

jdtsmith commented on activities.el:
```diff
@@ -801,6 +819,90 @@ activity's name is NAME."
+    (when-let ((activity (map-elt activities-activities name)))
+      (let (data (age-len 14))
+	(dolist (type '(last default))
+	  (let* ((func (intern (format "activities-activity-%s" type)))
```
Done.  Should also decide if we want to allow abbreviated ages (or user-customization of the age "words"), in which case padding will need to change.

alphapapa replied:
I don't think customizing that will be necessary.  If it were, there are about 5,000 other places in Emacsland that would be a higher priority.  :)

jdtsmith commented on activities.el:
```diff
@@ -801,6 +819,90 @@ activity's name is NAME."
+		 (time (map-elt (activities-activity-state-etc state) 'time))
+		 (window-state (activities-activity-state-window-state state))
+		 (buffers (window-state-buffers window-state))
+		 (files ( activities--map-window-state-leafs window-state
```
It's actually not extraneous: it prevents emacs-lisp-mode from indenting the subsequent line way over to the right under `window-state`.   It's a reasonably new feature.  But it is ugly, so I hacked it out.  How do you deal with this problem (I notice you often keep the 1st argument on the line with the function call)?

alphapapa replied:
Yes, but it's intended for, e.g. quoted plists, not for function calls.  I guess one could use it if one preferred, but it would be unidiomatic and visually confusing, as it would tend to obscure that it's a function call.

I don't like having one-symbol lines, so I avoid them when possible.  It's a balancing act between that and line length.

It's generally fine to fix such places manually, but in my projects I use `makem.sh` with its `lint-indentation` rule, which will give a warning about manually adjusted indentation, so please just indent it as Emacs automatically does and we'll live with it.

jdtsmith replied:
I do agree it's ugly, possibly uglier than one-symbol lines, but I tend to air on the compactness=readability side.  Restored the 1sl for now to be sure not to add linting errors.

jdtsmith commented:
BTW, are there other pieces of information inside an activity that would be relevant to show that occur to you?

alphapapa commented:
> BTW, are there other pieces of information inside an activity that would be relevant to show that occur to you?

I can't think of any.  For myself, the only one that matters is last-used time, anyway.  Let's keep this as simple as possible.

jdtsmith commented:
I could show both default and last ages, but that might be confusing.   

Along that line, do you think having the default and last buf+file cnt is useful or TMI?  Another approach would be to use some indicator (like `*`) that the state is "dirty" in the sense that its display buffer list has changed from the default.   So buffer cnt + file cnt + dirty flag + last mod age.

alphapapa commented:
> I could show both default and last ages, but that might be confusing.

I don't think that's necessary.  This is just for completing-read, not for "management."

> 
> Along that line, do you think having the default and last buf+file cnt is useful or TMI? Another approach would be to use some indicator (like `*`) that the state is "dirty" in the sense that its display buffer list has changed from the default. So buffer cnt + file cnt + dirty flag + last mod age.

For me, it's TMI.  I select activities by name, not by a number of how many windows or files are visible in it.  I can't think of a situation where I would need to know that.

However, knowing whether an activity has a last-used state could be useful, I think.

jdtsmith commented:
> > Along that line, do you think having the default and last buf+file cnt is useful or TMI? Another approach would be to use some indicator (like `*`) that the state is "dirty" 
> For me, it's TMI.

I do think I agree.  I like having the last buf/file count, since I use multiple "similar" activities, some richer than others (based on screen size).  But for the default counts, I'm just comparing to see if they changed at all (i.e. computing a "dirty" flag in my head).  For such a flag, what would you recommend?  I was thinking "same number of buffers and same buffer names = not dirty".

alphapapa commented:
I'm not sure what you mean by "dirty."  For me what would be interesting is whether the activity has a last-used state or just the default.

jdtsmith commented:
Sorry, I'm not sure I understand your idea.  If you have activity auto-saving on (which is the default), an activity will pretty much always have a "last" state, even if it's only lightly modified from the default (different window points, sizes, etc.).   So my idea of "dirty" is if the last state differs is an interesting way from the default — i.e. different buffers showing.  

I did need to accommodate a missing "last" state, but that's a pretty rare case for me, like when I've just defined a new state.

alphapapa commented:
> If you have activity auto-saving on (which is the default), an activity will pretty much always have a "last" state, even if it's only lightly modified from the default (different window points, sizes, etc.) ... I did need to accommodate a missing "last" state, but that's a pretty rare case for me, like when I've just defined a new state.

I often use `activities-kill`, which means that the activity won't have a last-used state.

> So my idea of "dirty" is if the last state differs is an interesting way from the default — i.e. different buffers showing.

I feel like that concept might be hard to communicate to the user, but if you want to add an asterisk or something and document it to mean that, I won't object.

jdtsmith commented:
OK sounds good; lack of asterisk would also mean "no last state".

One oddity I've found: when I `revert` an activity (which I use instead of kill), I'd expect its last state to be removed.  It isn't, and sometimes as a result there are two distinct sets of buffer names for the last and default state, even right after `revert` runs.  This happens if a buffer is a duplicate which has a `uniquify`d name.   I.e. last might have `main.c` while default has `main.c<proj>`.  Have you seen this?  Perhaps revert should just remove last, like kill does.

alphapapa commented:
> One oddity I've found: when I `revert` an activity (which I use instead of kill), I'd expect its last state to be removed. It isn't, and sometimes as a result there are two distinct sets of buffer names for the last and default state, even right after `revert` runs. This happens if a buffer is a duplicate which has a `uniquify`d name. I.e. last might have `main.c` while default has `main.c<proj>`. Have you seen this? Perhaps revert should just remove last, like kill does.

I can see pros and cons on either side.  Please feel free to open a new issue about that if you like.

jdtsmith commented:
Thanks.  I've pushed the `*` flag and a README update.  Let me know if you see anything else that needs addressing.

alphapapa commented on activities.el:
```diff
@@ -858,44 +859,44 @@ OLDEST-POSSIBLE is the oldest age in the `vc-annotate-color-map'."
+	      (setf (alist-get type data)
+		    (list (length (delq nil files))
+			  (and time (float-time (time-since time))) buffers)))))
+	(pcase-let* ((`(,last-file-cnt ,last-age ,last-bufs) (map-elt data 'last))
```
Forgive the nitpicking, but I'd still prefer not to abbreviate these variable names.  I'd suggest, e.g. `num-last-files`, `last-buffers`, etc.

jdtsmith replied:
Not at all!  You'll look at this more than me, so whatever names area easiest on your eyes is preferred.

alphapapa commented on activities.el:
```diff
@@ -858,44 +859,44 @@ OLDEST-POSSIBLE is the oldest age in the `vc-annotate-color-map'."
+		     (age (if last-age (min last-age default-age) default-age))
+		     (buf-cnt (length (or last-bufs default-bufs)))
+		     (file-cnt (or last-file-cnt default-file-cnt))
+		     (dirty (and last-bufs
```
In Emacs conventions, a flag variable like that should end in `p`, i.e. `dirtyp`.

Also, I prefer to use `when` rather than `and` here, as it clearly distinguishes the condition and the value forms.

alphapapa commented on activities.el:
```diff
@@ -858,44 +859,44 @@ OLDEST-POSSIBLE is the oldest age in the `vc-annotate-color-map'."
+		     (dirty (and last-bufs
+				 (or (/= (length last-bufs) (length default-bufs))
+				     (not (seq-set-equal-p last-bufs default-bufs)))))
+		     (ann (format "%s bufs %s files "
+				  (propertize (format "%2d" buf-cnt) 'face 'success)
+				  (propertize (format "%2d" file-cnt) 'face 'warning)))
```
Again, nitpicking, but I prefer, e.g. `annotation` to `ann`, and `num-buffers` to `buf-cnt`.

alphapapa reviewed (changes requested):
Thanks.  Also, could you post a final screenshot showing what it looks like with the asterisk?

jdtsmith commented on activities.el:
```diff
@@ -858,44 +859,44 @@ OLDEST-POSSIBLE is the oldest age in the `vc-annotate-color-map'."
+			       (map-nested-elt (cdr leaf)
+			 		       '(parameters activities-buffer))))))))
+	      (setf (alist-get type data)
+		    (list (length (delq nil files))
```
`window-state-leafs` actually just pushes the results of the function it was provided, which in this case involves `bookmark-get-filename`, which returns nil for non-file buffer bookmarks.   I think it's useful to preserve the nils in case you ever want to correlate the list against other calls.

jdtsmith commented:
Here's how it's looking (note the final `*` for modified activities):

<img width="573" alt="image" src="https://github.com/alphapapa/activities.el/assets/93749/28ecb0aa-1817-4c98-b3bf-713d71dce473">

Couple things I think still need a bit of thought: 

1. Active activities get saved automatically, which always updates their `time`.  I guess that could be considered OK, since it shows the activity being continuously saved, but when there are no changes to any of the activity's configuration, it is confusing that the activity age is updated.   I suppose we could replace the age annotation information with `<active>` for all active activities.
2. I had assumed activities were being sorted most recent first, since that's how mine had sorted.  But as the screenshot shows, this isn't the case.  Easy to do by pre-sorting the names based on recency.  Let me know if that should be a separate PR.

alphapapa commented:
Thanks for the screenshot.

> Couple things I think still need a bit of thought:
> 
>     1. Active activities get saved automatically, which always updates their `time`.  I guess that could be considered OK, since it shows the activity being continuously saved, but when there are no changes to any of the activity's configuration, it is confusing that the activity age is updated.   I suppose we could replace the age annotation information with `<active>` for all active activities.

Well, I do think that active activities should be marked in some way; it would seem like an obviously missing feature otherwise, IMO.

Replacing the age with some other string to indicate its being active would be one possibility.  OTOH the multiple colors already make that column a bit busy-looking.  I'm not sure what the best solution is.

We should probably also keep in mind potential window width in terms of the multiple columns of information.  Have you tested with a half-screen-width window?

>     2. I had assumed activities were being sorted most recent first, since that's how mine had sorted.  But as the screenshot shows, this isn't the case.  Easy to do by pre-sorting the names based on recency.  Let me know if that should be a separate PR.

I think a customizable sorting option should be part of this PR, if you're willing to work on it.  Having the age column but not sorting by it seems weird.  I'd guess that active ones should be sorted first, then the rest by age.

There's also https://github.com/alphapapa/frecency.el, which might be good to use.  But I think that library should be rewritten; I've learned a lot since I wrote it, and its API is a bit awkward, so I wouldn't suggest using it now.  But if the sorting function were customizeable, it could be used later.

jdtsmith commented:
> I do think that active activities should be marked in some way; it would seem like an obviously missing feature otherwise, IMO.

Yeah, agree.  Perhaps an `<A>` before the buffer count?  Or maybe better, change the codes up:

- `*`=active, modified
- `+`=inactive, modified
- `-`=active, unmodified 
- ` `=inactive, unmodified

> Have you tested with a half-screen-width window?

With the flex-spacer it's responsive to small widths: 

<img width="326" alt="image" src="https://github.com/alphapapa/activities.el/assets/93749/80c4a104-920c-451c-976d-9062856fe80b">

If we like it better, could measure the maximum activity width and pile everything up left (but still aligned).  I tend to prefer flush right.  

Will work on a simple sorting scheme.

jdtsmith commented:
Something else: `window-state-buffers` differ depending on whether that state is loaded and active, or not:

1. Buffers which get renamed by uniquify are updated in the `last` state to their `name<dir>` versions.
2. Buffers visiting symlinked files are named after the symlink (e.g. `.emacs`) in `default`, but the resolved file (e.g. `init.el`) in `last` (or maybe whichever state was loaded).

This leads to false-positive modified states.  Both of these mean I should check for equivalent _files_ (if available) or buffer names (if not).

jdtsmith commented:
OK I've got something I'm reasonably happy with, with `@` meaning active, `*` meaning modified.  I've also fixed the modification check to compare true filenames and then buffers as a backup, and it fixes the 2 corner cases mentioned above.  Here it is for the absurdly small window width of 45 chars:

<img width="333" alt="image" src="https://github.com/alphapapa/activities.el/assets/93749/9969ea86-a163-43ba-96cd-8f409c4b8ae1">

Obviously still need to sort, and couple of small structural bothers: 

1. When you revert, since last is not immediately updated or removed, the activity still shows up as modified for some time (see #86).  
2. As mentioned above, active activities have their timestamps continuously incremented, so they will _all_ mention the same 5 seconds age (or something small).  It would be better of course to update the timestamp (and data) for a given activity only if something actually changed in the window state.  I confirmed that `cl-tree-equal` works on window-states.

jdtsmith commented:
I got sort working, but I forgot that vertico substitutes it's own "most recent used then shortest candidate" sort function.  You can disable it by let-binding `vertico-sort-function nil` but that's a bit odd to throw in.  Thoughts?

jdtsmith commented:
Have you ever seen the persist file disappear?  It's happened to me twice in the past few days while hacking on this.  I note that persist removes the file if the value is ever the default value on save (`nil`).  I don't see any reason it would be getting set to nil, but perhaps edebug is involved.   Auto-save is on.

alphapapa commented:
> Have you ever seen the persist file disappear? It's happened to me twice in the past few days while hacking on this. I note that persist removes the file if the value is ever the default value on save (`nil`). I don't see any reason it would be getting set to nil, but perhaps edebug is involved. Auto-save is on.

It has happened to me with another package, but only when I used my `unpackaged/reload-package` command to reload the package's symbols after it was already loaded.  Since I stopped doing that, it hasn't happened again.  I'm guessing there's some kind of unexpected state and actions that can happen when developing if it involves reloading the file in which the variable is defined, or if `C-M-x` is used on the variable's definition.  So AFAICT it shouldn't ever affect users, but developers could find it quite inconvenient.  It would probably be worth filing an issue about and trying to nail it down and possibly fix it.

jdtsmith commented:
OK, see #87.  Give the sort and active marker a look and let me know what you think about preventing the time-tick from updating unless "things have changed".

jdtsmith commented:
Anything further needed here?  If you like I can look into updating activity timestamp on save only if buffers changed.

alphapapa commented:
I have a lot going on right now, so it may be a while before I can review this further.  I appreciate your patience.

jdtsmith commented:
No rush on review, but I wanted to mention that I factored out the buffer/file comparison code from the annotation function, and  updated `activities-save` using it to retain the timestamp if the list of buffers/files has not changed.  This works _much_ better with the idle-time `save-all`, no longer spuriously updating the age of active but "idle" activities.   Working quite well so far.

jdtsmith commented:
Gentle ping on this one.  Been in active use for quite some time without any issues on my end.

alphapapa commented:
> Gentle ping on this one. Been in active use for quite some time without any issues on my end.

Thanks.  My to-do list is quite long and my time has been limited.  I guess this can be merged now...

alphapapa commented on activities.el:
```diff
@@ -801,6 +851,116 @@ activity's name is NAME."
+
+(defun activities--age (age &optional abbreviate)
+  "Summarize AGE and possibly ABBREVIATE.
+Adapted from `magit--age'."
```
I'm not sure if this can be adapted here, because Activities is in GNU ELPA and is copyrighted by the FSF, while Magit is not copyrighted by the FSF.  @tarsius Could you please advise here?

tarsius replied:
That's correct. However, I am the only one who ever touched this function and the associated variable, and would be happy to contribute this to Emacs. I would ask @jdtsmith to do the work of getting it integrated (or, if that's what the maintainers request, replaced by a better implementation).

alphapapa replied:
@tarsius Thanks.  With your being the author, if you consent, then, can it just be "contributed to Emacs" via this package, since ELPA packages are "part of GNU Emacs"?

tarsius replied:
Well yes, but let's wait and see if @jdtsmith would be willing to get it into Emacs first. Doesn't have to be this implementation, but something like this would certainly be useful in more than these two packages.

As for the used algorithm, one might want to consider using another threshold than 1. I don't really mind that 2.0 and 2.49 years are both "2 years", but when it comes to 0.51 and 1.49 both being "1 year", I feel a bit differently. Maybe this should be customizable, maybe even per time unit.

If I remember correctly, I did shop around for other implementations before writing mine, but the once I found all addressed this by using more than one time unit, which doesn't work in our limited space constraint.

jdtsmith replied:
Happy to help out.  What did you envision?  I also use some color and date-related functionality from `vc-annotate`, perhaps it could go there (though I rarely use `vc` anymore, because magit ;).  

A little more granularity would be nice; maybe a unicode ½ could enter.

tarsius replied:
> A little more granularity would be nice; maybe a unicode ½ could enter.

My 2 cents: Sure ½ years is nice, but 12½ years is unnecessary "precision". I like this progression of half year steps better 6 months, 12 months, 18 months, 2 years, 3 years (for 2.5), 3 years (for 3), and so on. This is actually more precise (in the range were it matters) than the ½ approach: 4/6/8 months vs. ½/½/½ years. And that's what I don't like about the half-step approach, it suggests precision that just isn't there.

jdtsmith replied:
The progression would go 1 hour, 1½ hours, 1 day, 1½ days, 1 week, 1½ weeks, 1 month, 1½ months, 6 months, 1 year, 1½ years.  I can see an argument for dropping the ½ and reverting to integer rounding after a certain base value, perhaps 3½?  What about 7½ months vs. 7 or 8 months?  I could go either way on that one.  Most relevant for years (I briefly toyed with adding "decades" but... no).  I agree that 29½ years is fairly silly.   The current `seconds-to-string` uses 1/100ths across the board btw.

tarsius replied:
https://debbugs.gnu.org/cgi/bugreport.cgi?bug=71572

tarsius replied:
I think I am going to stick what I have now.

> The current seconds-to-string uses 1/100ths across the board btw.

Had I known that function and if it used 1/10, I probably would have used it.

A code comment mentions that there is a fancier implementation to be found in `mastodon.el`. I wasn't able to locate that, could you?

jdtsmith replied:
No I didn't.  Take a look at the patch I submitted (should also be in your inbox).  Happy for further  input.

tarsius replied:
There obviously are different approaches to this task. For the reasons given, I don't like the half-step approach much, and looking at the implementation did not change that. That shouldn't keep you from iterating on it. There's room for different approaches. If something lands in Emacs, it would probably be a good idea, if it were done in a way that allowed for switching in an alternative implementation. (I.e., using a `...-function` variable, though in the absence of that, an `:override` advice would of course also work.)

jdtsmith replied:
A good design goal IMO is "something tarsius would be happy to sub in for magit--age at some point", so feel free to express your ideas [there](https://debbugs.gnu.org/cgi/bugreport.cgi?bug=71573).  So far no response.  `HALF` is optional btw, default off.

alphapapa commented on activities.el:
```diff
@@ -503,6 +514,39 @@ To be called from `kill-emacs-hook'."
+(defun activities--buffers-and-files (state)
+  "Return a list of buffers and files from STATE.
+STATE is a window-state.  The returned list contains elements of
+form (BUFFER . FILE) assocaited with the activity."
```
Nitpick: Spelling typo.  :)

alphapapa commented on activities.el:
```diff
@@ -503,6 +514,39 @@ To be called from `kill-emacs-hook'."
+(defun activities--map-window-state-leafs (state func)
+  "Call FUNC on all the leaf nodes in window-state STATE.
+The resulting values are returned as a list."
+  (let (ret)
```
Nitpick: Could we call this variable `values` instead of `ret`?

alphapapa commented on activities.el:
```diff
@@ -503,6 +514,39 @@ To be called from `kill-emacs-hook'."
 ;;;; Functions
 
+(defun activities--map-window-state-leafs (state func)
+  "Call FUNC on all the leaf nodes in window-state STATE.
```
I prefer for the first line to say what the function returns.  How about something like, "Return a list of values of having applied FUNC to leaf nodes in STATE."?  Maybe the function could be called `activities--mapcar-window-state-leafs` to additionally imply that it returns a list like `mapcar`.  What do you think?

alphapapa commented on activities.el:
```diff
@@ -503,6 +514,39 @@ To be called from `kill-emacs-hook'."
+  (activities--map-window-state-leafs
+   (activities-activity-state-window-state state)
+   (lambda (leaf)
+     (let ((rec (map-nested-elt (cdr leaf) '(parameters activities-buffer))))
```
Could we use a more descriptive variable name than `rec` here?

alphapapa commented on activities.el:
```diff
@@ -503,6 +514,39 @@ To be called from `kill-emacs-hook'."
+       (cons (activities-buffer-name rec) (activities-buffer-filename rec))))))
+
+(defun activities--buffers-and-files-differ-p (bf1 bf2)
+  "Return t if BF1 and BF2 are not the same set of files or buffers."
```
This should probably say "Return non-nil if...", because it shouldn't matter whether it's `t` or another non-nil value.

Also, it would be good to explain what BF1 and BF2 are (checkdoc will complain otherwise, at least).

alphapapa commented on activities.el:
```diff
@@ -503,6 +514,39 @@ To be called from `kill-emacs-hook'."
+(defun activities--buffers-and-files-differ-p (bf1 bf2)
+  "Return t if BF1 and BF2 are not the same set of files or buffers."
+  (cl-labels ((file-or-buffer (cell)
+		"Given (buffer . file), return the true filename or (if none) buffer."
```
I like having the docstring in the label function, but it should probably follow the convention if it's to be there.

alphapapa commented on activities.el:
```diff
@@ -503,6 +514,39 @@ To be called from `kill-emacs-hook'."
+  "Return t if BF1 and BF2 are not the same set of files or buffers."
+  (cl-labels ((file-or-buffer (cell)
+		"Given (buffer . file), return the true filename or (if none) buffer."
+		(if (cdr cell) (file-truename (cdr cell)) (car cell))))
```
Should we use `abbreviate-file-name` as well?

jdtsmith replied:
These files (or buffers) are used exclusively to determine is an activity has changed, so abbreviating the files wouldn't add anything I don't think.  They don't get displayed.

alphapapa commented on activities.el:
```diff
@@ -801,6 +851,116 @@ activity's name is NAME."
+    (?h "hour"   "hours"   ,(* 60 60))
+    (?m "minute" "minutes" 60)
+    (?s "second" "seconds" 1))
+  "Age specification.  See `magit--age-spec', which this duplicates.")
```
See copyright question below.

jdtsmith replied:
I think the right approach is to get something like this into time-date.el, copy it here temporarily, and once it shows up in `compat`, remove it.  I'll send in a patch.

jdtsmith replied:
One issue is this similar existing function:

```elisp
(defvar seconds-to-string
  (list (list 1 "ms" 0.001)
        (list 100 "s" 1)
        (list (* 60 100) "m" 60.0)
        (list (* 3600 30) "h" 3600.0)
        (list (* 3600 24 400) "d" (* 3600.0 24.0))
        (list nil "y" (* 365.25 24 3600)))
  "Formatting used by the function `seconds-to-string'.")
;;;###autoload
(defun seconds-to-string (delay)
  ;; FIXME: There's a similar (tho fancier) function in mastodon.el!
  "Convert the time interval in seconds to a short string."
  (cond ((> 0 delay) (concat "-" (seconds-to-string (- delay))))
        ((= 0 delay) "0s")
        (t (let ((sts seconds-to-string) here)
             (while (and (car (setq here (pop sts)))
                         (<= (car here) delay)))
             (concat (format "%.2f" (/ delay (car (cddr here)))) (cadr here))))))
```

jdtsmith replied:
See bug#71573.

alphapapa replied:
FWIW, I replied to that bug report linking to this code in my `ts` library: https://github.com/alphapapa/ts.el/blob/master/ts.el#L440-L491  Although it works a bit differently, I've used it extensively in some of my other packages, and profiling shows that it performs very well in terms of runtime and GC.  If any of it seems useful, I'll be glad to contribute it upstream as well (I'd be glad to contribute all of `ts`, in fact, but I haven't exactly proposed that yet).

jdtsmith replied:
Thanks.  My guess is the attention to cutting v30 will mean this sits for a while.  I tried to adapt to the style of date-time but am not committed to any particular version. I would like to include weeks and months; to me that's the key elaboration of `magit--age`
(plus the modest resolution).  

How about we give it a week and if no movement, put in a version ourselves, with the aim to remove it once the patch is included in a release (+compat)?

jdtsmith replied:
Some uptake on the patch adding a new `readable` option to `seconds-to-string`, but Eli just confirmed it wouldn't land until Emacs 31, so I propose we include a copy here along the same lines, with a note to replace it eventually.  Pushed changes along these lines.

alphapapa commented on activities.el:
```diff
@@ -801,6 +851,116 @@ activity's name is NAME."
+							     (length age-annotation)))))
+		  annotation age-annotation dirty-annotation))))))
+
+(defun activities-sort-by-active-age (names)
```
If this function is only used in one place, I'd like to avoid defining it as a top-level symbol.

jdtsmith replied:
There is a customization option `activities-sort-function` for which this function is the default value.   This allows the user to specifying another sort.

alphapapa commented on activities.el:
```diff
@@ -801,6 +851,116 @@ activity's name is NAME."
+		 (activep-b (activities-activity-active-p activity-b)))
+	    (or (and activep-a (not activep-b)) (time-less-p time-b time-a))))))
+
+(defun activities--completion-table (activities)
```
Do you foresee this function being called elsewhere than in `activities-completing-read`?  If not, I think I'd prefer for it to be defined inside that function, to avoid the top-level symbol.

jdtsmith replied:
Done.

alphapapa commented on activities.el:
```diff
@@ -801,6 +851,116 @@ activity's name is NAME."
+	    for etc = (activities-activity-state-etc state)
+	    maximize (float-time (time-since (map-elt etc 'time))))))
+
+(defun activities--annotate (max-age oldest-possible)
```
Forgive my laziness, since my time is limited, but could this be done without defining this top-level symbol?  I really like to try to avoid defining new ones.

jdtsmith replied:
I could shove this into `activities-completing-read` just as I did the completion table, but it's a pretty meaty function.  What do you think?

jdtsmith replied:
I managed to squeeze the annotation function in there.

alphapapa commented on activities.el:
```diff
@@ -66,6 +66,7 @@
 (require 'map)
 (require 'persist)
 (require 'subr-x)
+(require 'vc-annotate)
```
OT1H this seems like an elegant reuse of code; OTOH many users would never load this library otherwise (myself being one of them).  Is there another way we could do the equivalent, while still reusing some face/color definitions?

jdtsmith replied:
I've used them before in other packages.  In fact I know about these colors because a better color-age dynamic range mapping was the first contribution I made to emacs (in a prior millennium).  We could rip the colors out of there and duplicate, but that colormap has been around long enough that I wonder if anyone has customized it.  It is used for blame functionality, so quite similar in spirit.  

It's trivial to select a precise color based on fractional age, using `(color-hsl-to-rgb)`, perhaps also giving users the ability to pick a lightness/saturation value in a defcustom.   Thoughts?

E.g. 
<img width="386" alt="image" src="https://github.com/alphapapa/activities.el/assets/93749/abcbcb8f-e410-450e-9e33-4b0610a7dac2">

alphapapa replied:
I'd suggest perhaps using this function I wrote for `ement`, which does something very similar but a bit more nuanced: https://github.com/alphapapa/ement.el/blob/9f6cbb8b406160c950d8932d66f3629aba091ec8/ement-room-list.el#L849-L890

jdtsmith replied:
I eliminated vc-annotate, and went with a simple blend from a specified old to new color, and a blend fraction with the default foreground.

alphapapa reviewed (changes requested):
Well, apologies for the long delay, and then asking for more changes.  Thanks for your work on this.  It will be merged in the end.  :)

jdtsmith commented:
> > Gentle ping on this one. Been in active use for quite some time without any issues on my end.
> 
> Thanks. My to-do list is quite long and my time has been limited. I guess this can be merged now...

Happy to move at your pace, no problem.  I tend to forget such things myself and so appreciate a ping here and there.

alphapapa commented on activities.el:
```diff
@@ -332,6 +333,26 @@ Kills buffers that have only been shown in that activity's
 frame/tab."
   :type 'boolean)
 
+(defcustom activities-sort-function #'activities-sort-by-active-age
+  "A function to use to sort by when prompting for activities.
+If nil, no sorting will be applied.  The function should take two
+arguments, both activity names (strings).  It should return
+non-nil if the first activity should sort before the second.  By
+default, a function is used which sorts active activities first,
+and then by age since modification."
+  :type '(choice (const :tag "No sorting" nil)
+		 (function :tag "A specific function")))
```
I think this could be a bit more idiomatic by describing the option in terms of the result rather than the predicate used to achieve it.  For example, naming the option `activities-sort-by`.  Then the type spec should include the default predicate, and the tag would be, e.g. "Age", so the user could choose to sort by "Age" or "custom predicate."

The docstring needn't describe the predicate beyond its arguments, I think, since that's just how a sort predicate works.

alphapapa commented on activities.el:
```diff
@@ -332,6 +333,26 @@ Kills buffers that have only been shown in that activity's
+  :type '(choice (const :tag "No sorting" nil)
+		 (function :tag "A specific function")))
+
+(defcustom activities-annotation-colors '("blue" "red" 0.65)
+  "Colors to use for annotating activity age.
+A list (OLD-COLOR NEW-COLOR BLEND-FRAC).  These are used during
+activity selection, with the activity color chosen between
+OLD-COLOR and NEW-COLOR, based on the activity's age.  This color
+is blended into the default foreground with a fraction
+BLEND-FRAC, and used to display the activity age."
+  :type '(list (color :tag "Old Color") (color :tag "New Color")
+	       (float :tag "Blend Fraction")))
```
It might be more idiomatic to call the float the "alpha" or "opacity".  Also, it would be good to make the docstring a bit more concise.

alphapapa commented on activities.el:
```diff
@@ -503,6 +524,45 @@ To be called from `kill-emacs-hook'."
+       (cons (activities-buffer-name buffer-rec)
+	     (activities-buffer-filename buffer-rec))))))
+
+(defun activities--buffers-and-files-differ-p (bf1 bf2)
+  "Return non-nil if BF1 and BF2 are not the same set of files or buffers.
+Each of BF1 and BF2 is a list of buffer and files, as returned
+from `activities--buffers-and-files'."
+  (cl-labels ((file-or-buffer (cell)
+		"Given a CELL, return the true filename or buffer.
+The CELL is a (BUFFER . FILE) cons.  If the file is nil, BUFFER is returned."
+		(if (cdr cell) (file-truename (cdr cell)) (car cell))))
+    (not (seq-set-equal-p (mapcar #'file-or-buffer bf1)
+			  (mapcar #'file-or-buffer bf2)))))
```
Very nitpicky, but I'd prefer to name these arguments, e.g. `bf-a` and `bf-b`, because they aren't sequential in nature.  (Maybe just an idiosyncrasy of how my mind works.)

alphapapa commented on activities.el:
```diff
@@ -801,6 +867,63 @@ activity's name is NAME."
+  "Age specification.  See `magit--age-spec', which this duplicates.")
+
+(defun activities--age (age &optional abbrev)
+  "Summarize AGE.
+Abbreviate the units if ABBREV is non-nil.  Based on
+`magit--age'."
```
It's important to mention the provenance of the code, but that can go in a comment rather than the docstring.

alphapapa commented on activities.el:
```diff
@@ -801,6 +867,63 @@ activity's name is NAME."
+  "Summarize AGE.
+Abbreviate the units if ABBREV is non-nil.  Based on
+`magit--age'."
+  ;; TODO: replace this if seconds-to-string adds READABLE support
```
Does this TODO need to be updated?  Also, a link to the relevant Emacs bug should be included.

jdtsmith replied:
Not sure if you are following that bug (#[71572](https://yhetil.org/emacs-bugs/D91AFAEE-C909-4238-AFF2-226727D4B0F6@gmail.com/T/#t)), but some expressed concern about including a unicode char.  I don't personally think that's an issue in this day and age, but wanted to check.

alphapapa replied:
I've been following it but not very closely.  I didn't realize how much of a bikeshed it was going to become (not a hostile one--I'm just surprised at how long it's gone on).  :)  Maybe that's why there are several implementations floating around, because everyone has their own preferences.

I don't have a problem with using Unicode here, as long as the character is readable, and not only present in very recent versions of common fonts.

jdtsmith replied:
I fully expected the bike-shedding ;).  We'll see how it converges; it appears it's a problem that's been solved many times.  As long as something reasonably flexible appears in Emacs 31 it probably doesn't matter much.  If it goes in without the ½, which seems likely, we can always give precision = 0.5 and modify `x.5` to read `x½`.

alphapapa commented on activities.el:
```diff
@@ -801,6 +867,63 @@ activity's name is NAME."
+  "Return the list activity NAMES sorted active first, then by age."
+  (sort names
+	(lambda (a b)
+	  (let* ((activity-a (map-elt activities-activities a))
+		 (state-a (or (activities-activity-last activity-a)
+			      (activities-activity-default activity-a)))
+		 (time-a (map-elt (activities-activity-state-etc state-a) 'time))
+		 (activep-a (activities-activity-active-p activity-a))
+		 (activity-b (map-elt activities-activities b))
+		 (state-b (or (activities-activity-last activity-b)
+			      (activities-activity-default activity-b)))
+		 (time-b (map-elt (activities-activity-state-etc state-b) 'time))
+		 (activep-b (activities-activity-active-p activity-b)))
```
Would you be interested in using `pcase-let*` to do that binding?  :)  If not, that's okay; just an idea.

jdtsmith replied:
Sure see how that looks.  I went for a `cl-labels` an `pcase-lambda` to try to make it a bit more readable.  Eye of the beholder of course (and many stalwart elispers think pcase is de-facto unreadable).

alphapapa commented on activities.el:
```diff
@@ -801,6 +867,63 @@ activity's name is NAME."
+			      (activities-activity-default activity-b)))
+		 (time-b (map-elt (activities-activity-state-etc state-b) 'time))
+		 (activep-b (activities-activity-active-p activity-b)))
+	    (or (and activep-a (not activep-b)) (time-less-p time-b time-a))))))
```
This would be a bit easier to read if the two OR'ed forms were on separate lines.

alphapapa commented on activities.el:
```diff
@@ -809,9 +932,60 @@ activity's name is NAME."
+      ((names (activities-names activities))
+       (max-age (activities--oldest-age activities))
+       (`(,old-col ,new-col ,blend-frac) activities-annotation-colors)
+       (annotation-function
```
FWIW, I'll probably move the annotation function into a `cl-labels` form; just a preference of mine.

jdtsmith replied:
Honestly it does feel we are packing a _lot_ into this function body.  I'd certainly reach for an internal function to keep it clearer, but again what matters is what's clear to _you_.  I converted them both to named local functions with `cl-labels`.  I didn't think I'd like it, but I do like that it's explicit that they are functions now in the call to `completing-read`, and have docstrings for clarity.

alphapapa commented on README.org:
```diff
@@ -135,6 +135,15 @@ Key bindings are, as always, ultimately up to the user.  However, in [[Configura
 
 When option ~activities-bookmark-store~ is enabled, an Emacs bookmark is stored when a new activity is made.  This allows the command ~bookmark-jump~ (~C-x r b~) to be used to resume an activity (helping to universalize the bookmark system).
 
+** Annotation and Sort
```
This section should probably be titled "Completion".

alphapapa commented on README.org:
```diff
@@ -135,6 +135,15 @@ Key bindings are, as always, ultimately up to the user.  However, in [[Configura
 
+** Annotation and Sort
+
+When selecting an activity in the mini-buffer, annotations are sorted active first, then by most recent modification age.  If you are using an interface that supports /annotations/, additional information is shown alongside the activity name:
+
+- an `@` symbol, for active activities,
+- the number of buffers and files the activity contains,
+- a color-coded age since the activity was last updated, and
+- a final ~*~ if the activity's list of buffers and files has been modified from its default state.
```
I think the sorting order will be obvious to the user, so probably the only thing that needs to be documented is what the non-obvious symbols mean.

alphapapa reviewed (changes requested):
Forgive the additional round of review and comments.  If you don't have enough "steam" to make some of the stylistic changes, I can make them myself before merging.

jdtsmith commented:
> Forgive the additional round of review and comments. If you don't have enough "steam" to make some of the stylistic changes, I can make them myself before merging.

No worries at all.  In my view, asking an author to incorporate significant changes, and (far more importantly) _support them for years or decades to come_ calls for real patience.  As long as good faith progress is being made towards merge, I'm happy to iterate.

alphapapa commented:
> > Forgive the additional round of review and comments. If you don't have enough "steam" to make some of the stylistic changes, I can make them myself before merging.
> 
> No worries at all. In my view, asking an author to incorporate significant changes, and (far more importantly) _support them for years or decades to come_ calls for real patience. As long as good faith progress is being made towards merge, I'm happy to iterate.

I appreciate that.  Sometimes I wonder if I should handle some nitpicks myself before merging; I try to consider whether mentioning it will serve any educational purpose, but then I don't always know how familiar the contributor already is, and whether they just have other preferences.

jdtsmith commented:
> > > Forgive the additional round of review and comments. If you don't have enough "steam" to make some of the stylistic changes, I can make them myself before merging.
> > 
> > 
> > No worries at all. In my view, asking an author to incorporate significant changes, and (far more importantly) _support them for years or decades to come_ calls for real patience. As long as good faith progress is being made towards merge, I'm happy to iterate.
> 
> I appreciate that. Sometimes I wonder if I should handle some nitpicks myself before merging; 

I'd say feel free to push any changes yourself if they are glaring and not worth describing.  But I appreciate the collaborative feel your approach creates.  I do a lot of work on other people's packages I admire, and some very capable authors have trouble with that, and can't resist NIH syndrome.  So I'd say your approach is very valuable, especially to any less experienced contributors who come along.

> I try to consider whether mentioning it will serve any educational purpose, but then I don't always know how familiar the contributor already is, and whether they just have other preferences.

I do have some different preferences but I always enjoy learning how others approach a problem.  And I don't mind adapting to the local accent (Canadian ? ;).  So some of both here.

As an example I mentioned above, I don't often reach for `cl-labels` when composing `lambda`'s would do — i.e. when you are just going to pass the function on as an argument, once.  But I found to my surprise I like quite the readability and mimetics (of `some-func` vs. `#'my-local-func`), so may use that more.

Other than deciding whether we are OK with ½ in the ages I think this is ready from my end.  Feel free to season to taste.

jdtsmith commented:
Actually one more thing: I don't like that `max-age` maximizes across both `last` and `default.`  It should use `last` _or_ `default.`

alphapapa commented:
> I do have some different preferences but I always enjoy learning how others approach a problem. And I don't mind adapting to the local accent (Canadian ? ;). So some of both here.

Do I seem Canadian?  I'm American, born and raised.  :)

`C-h f pcase RET` shows me:
```
-- (cl-struct TYPE &rest FIELDS)

Pcase patterns that match cl-struct EXPVAL of type TYPE.
Elements of FIELDS can be of the form (NAME PAT) in which case the
contents of field NAME is matched against PAT, or they can be of
the form NAME which is a shorthand for (NAME NAME).
```
But that might only show up after loading the relevant library that defines that `pcase` pattern, IIRC.

Also, FYI:
```
-- (cl-type TYPE)

Pcase pattern that matches objects of TYPE.
TYPE is a type descriptor as accepted by ‘cl-typep’, which see.
```

jdtsmith commented:
> `C-h f pcase RET` shows me:
> 
> ```
> -- (cl-struct TYPE &rest FIELDS)
> 
> Pcase patterns that match cl-struct EXPVAL of type TYPE.
> Elements of FIELDS can be of the form (NAME PAT) in which case the
> contents of field NAME is matched against PAT, or they can be of
> the form NAME which is a shorthand for (NAME NAME).
> ```

Ahh, right you are.  What's strange is there's nothing in `info`, my usual go-to for `pcase.`  In general the info and docstring are similar enough that it's strange what is omitted.  For example, both include `cl-type`, but only the docstring mentions `seq`, `radix-tree-leaf` and `cl-struct`.

alphapapa commented:
> Ahh, right you are. What's strange is there's nothing in `info`, my usual go-to for `pcase.` In general the info and docstring are similar enough that it's strange what is omitted. For example, both include `cl-type`, but only the docstring mentions `seq`, `radix-tree-leaf` and `cl-struct`.

Probably an oversight, then, and probably worth a bug report.

alphapapa commented on activities.el:
```diff
@@ -332,6 +333,26 @@ Kills buffers that have only been shown in that activity's
+take two arguments, both activity names (strings), and return
+non-nil if the first activity should sort before the second."
+  :type `(choice (const :tag "No sorting" nil)
+		 (const :tag "Active state and age" ,#'activities-sort-by-active-age)
```
Rather than `const`, this should probably be `function-item`.  (The details of the customization system are covered in the Elisp manual; there's a lot to it.)

jdtsmith replied:
OK done.  I dislike how `function-item` and others mention the function name in the custom interface, instead of the `:tag` so that's why I used `const`.  But adding `:doc` is a decent workaround.

alphapapa replied:
Hm, that's a good point.  I'll leave this decision up to you, then.  :)

jdtsmith replied:
I think it's fine with `:doc`, and certainly more semantically meaningful.

alphapapa commented on activities.el:
```diff
@@ -809,11 +929,65 @@ activity's name is NAME."
+	   (if (eq action 'metadata)
+	       `(metadata (annotation-function . ,#'activity-annotation-function)
+			  ,@(when activities-sort-by
+			      `(,(cons 'display-sort-function activities-sort-by))))
```
Is it necessary to support sorting being optional?  It seems like disabling sorting would be of no benefit.

jdtsmith replied:
I guess no-sorting is actually "sorting by the order I made these activities", right?  Could be useful to someone.

alphapapa replied:
I don't think the order of the `activities-activities` list is defined, so "no sorting" would be equivalent to "unknown order," which doesn't seem useful to me.  Unless there's a potential performance issue that would make it necessary to disable sorting, I think I'd rather go for the simpler code that expects the sort-by function to always be set (it could be set to `always` or `ignore` to effectively disable sorting, if the user really wanted).

jdtsmith replied:
Fair enough, the user can always slot `#'ignore` in if they want no change.  I also found a last minute slight flaw in the sort logic and fixed that.

alphapapa reviewed (changes requested):
Thanks, I think this will be ready to merge after this.

alphapapa commented:
Thanks for your considerable patience in this process.  One final thing: would you be willing to write a changelog entry for this branch's changes?

jdtsmith commented:
> Thanks for your considerable patience in this process. One final thing: would you be willing to write a changelog entry for this branch's changes?

No problem, done.

jdtsmith commented:
Anything more to do on this PR, @alphapapa?

alphapapa commented:
> Anything more to do on this PR, @alphapapa?

Thanks for your patience.

By the way, forgive me for asking about this again, but I can't find where I asked you specifically before, and this contribution is larger, and I'm trying to be extra diligent about this when accepting contributions: You have completed the FSF copyright assignment for Emacs, right?  I see in emacs.git some contributions from you, but it's not clear whether they are covered by an assignment (e.g. the recent ones under jdtsmith@gmail.com say they are exempt).  

To be certain, I'm required by the Emacs maintainers to ask them for confirmation for a given name and email address.  So is the right name and email address to ask them about `"JD Smith <jdtsmith@gmail.com>"`?  (Again, apologies if this is redundant, but I'm trying to be very careful.)

jdtsmith commented:
I appreciate your diligence (which no doubt saves headaches down the road).  Yes, I completed this 23 yrs ago ;).  I recently updated my email with the FSF, so that should work for confirmation.

alphapapa commented:
Thanks!  Email sent.  Should hear back soon, then I can merge this PR.

alphapapa commented:
Thanks for your patience.  Eli confirmed your CA (a while ago, I just didn't get around to handling the message until now).

jdtsmith commented:
Making my list and checking it twice... I see this one is still missing a checkmark ;).

alphapapa commented:
@jdtsmith Thanks, you've been extremely patient.  I'll merge this soon... (even writing a note on paper to remind me)

alphapapa commented:
@jdtsmith I rebased on master in preparation for merging.  

While doing some final checks, I noticed one final thing to ask about: While your FSF CA was done as `jdtsmith@gmail.com`, these git commits have the address `93749+jdtsmith@users.noreply.github.com`.  Is that intentional?  It would be more convenient for copyright attribution if the same address were used for authoring the commits as used when filing copyright assignment.  Otherwise, if someone goes auditing in the future, one would have to dig up these conversations and try to find the link in the chain between that "noreply.github.com" address and your real address which is on file.

jdtsmith commented:
I have the "keep email address private" option selected, which means github refuses pushes with your actual email as author:

```
remote: error: GH007: Your push would publish a private email address.        
remote: You can make your email public or disable this protection by visiting:        
remote: https://github.com/settings/emails        
```


But since it's been mentioned here I temporarily changed that and force-pushed with updated author email.  It was a lot of commits!  Luckily magit with rebase-interactive and a quick keyboard macro made quick work of it.

jdtsmith commented:
BTW, RE our discussions above of richer abbreviated dates, my related patch was [merged to master](https://debbugs.gnu.org/cgi/bugreport.cgi?bug=71572) recently, and will soon be added to `compat` (which I notice you do not use?).  For the future.

alphapapa commented:
> But since it's been mentioned here I temporarily changed that and force-pushed with updated author email. It was a lot of commits! Luckily magit with rebase-interactive and a quick keyboard macro made quick work of it.

Thanks.

So, after about 8 months, this is finally merged.  Thanks for your contribution and, of course, your patience.

> BTW, RE our discussions above of richer abbreviated dates, my related patch was [merged to master](https://debbugs.gnu.org/cgi/bugreport.cgi?bug=71572) recently, and will soon be added to compat (which I notice you do not use?). For the future.

I use `compat` in my packages where it's needed.  There hasn't been a need in this one yet.  If it helps, or allows removing code, by all means, let's use it.
